lib/hmmlearn/tests/test_gmm_hmm_multisequence.py

```diff
@@ -0,0 +1,290 @@
+                [0.00313668, 0.47089654, 0.55917096, 0.37248307],
+                [0.00320294, 0.45442477, 0.57189673, 0.38878405],
+                [0.00317113, 0.43952465, 0.59469926, 0.3535012],
+                [0.00323172, 0.46230638, 0.5713085, 0.39267808],
+                [0.00323527, 0.50055844, 0.52773774, 0.4445253],
+                [0.00322912, 0.5141386, 0.501789, 0.488796],
+                [0.00312394, 0.53217465, 0.47047156, 0.49383518],
+                [0.00305265, 0.5208433, 0.50014937, 0.459741],
+                [0.00305853, 0.51386, 0.5120822, 0.43915167],
+                [0.00311382, 0.47278965, 0.57036394, 0.45423037],
+                [0.00318763, 0.46807718, 0.5674287, 0.43843716],
+                [0.00318154, 0.47042188, 0.5779207, 0.44580007],
+                [0.00303793, 0.46571356, 0.5818657, 0.4107192],
+                [0.00299003, 0.44406536, 0.60040486, 0.38005733],
+                [0.00296462, 0.44092676, 0.59199274, 0.3560812],
+                [0.00307418, 0.45244223, 0.56540257, 0.33132356],
+                [0.00302929, 0.47221994, 0.5470938, 0.36822143],
+                [0.00301338, 0.47062206, 0.54911035, 0.3680197],
+                [0.00291349, 0.45564204, 0.5619591, 0.3808152],
+                [0.00295025, 0.47307628, 0.54229873, 0.35219967],
+                [0.00289911, 0.4725018, 0.54288876, 0.38314143],
+                [0.00291358, 0.45272404, 0.5728319, 0.37214962],
+                [0.00294044, 0.4570347, 0.5649049, 0.42110214],
+                [0.00298315, 0.4561437, 0.55468136, 0.40142712],
+                [0.00303691, 0.49613172, 0.52464443, 0.42899716],
+                [0.00300609, 0.49105296, 0.5310325, 0.41341364],
+                [0.00305483, 0.532239, 0.485052, 0.46893936],
+                [0.00298801, 0.51202387, 0.49897632, 0.45285705],
+                [0.0030795, 0.5070261, 0.5062058, 0.44057184],
+                [0.00303104, 0.47002167, 0.56080073, 0.40074068],
+                [0.00308184, 0.4575324, 0.56803167, 0.3928257],
+                [0.00291516, 0.44490653, 0.58391964, 0.40129626],
+                [0.00289649, 0.4531514, 0.5818511, 0.407385],
+                [0.00284487, 0.4488143, 0.5850243, 0.41675568],
+                [0.00291576, 0.4616304, 0.57469726, 0.40793785],
+                [0.00287342, 0.46473294, 0.54888374, 0.41002542],
+                [0.00303153, 0.4966541, 0.5055506, 0.4316879],
+                [0.00327755, 0.45762977, 0.529044, 0.4493881],
+            ],
+            dtype=np.float32,
+        ),
+    ]
+)
+
+
+class FailedToConverge(RuntimeError):
+    pass
+
+
+class StrictMonitor(ConvergenceMonitor):
+    @property
+    def converged(self):
+        if len(self.history) < 2:
+            return False
+
+        if np.isnan(self.history[-1]):
+            raise FailedToConverge("logprob must not be nan")
+
+        if self.history[-1] < self.history[-2] - self.tol:
+            raise FailedToConverge("logprob must be non-decreasing")
```

Comment:
I would just return False in these two cases for the purpose of making test implementation simpler (or you can use `assert`s and rely on pytest converting assert failures to test failures).

Reply:
When the monitor `converged` property returns `False` the calling `fit` method of `_BaseHMM` interprets that as "we haven't converged yet, let's keep iterating" until max iters are exhausted. The monitor doesn't seem to have a defined way to signal to `_BaseHMM` that something has gone badly wrong and the fit method abort. 

If we sense nans or decreasing logprod during iterations it seems more helpful (to maintainer trying to understand cause of test failure from symptoms) to immediately abort at first sign something is wrong instead of keeping on running.  So of the two alternatives you propose, using assertions seems superior to me than returning False, as that will abort the fit.

Reply:
switching to use asserts here produces arguably superior logs when the test fails: instead of "assert all [ false false false false ]" it fails quicker and pytest dumps the entire strict monitor state that shows the logprob history values decreasing.

i've removed the exception handling. remains of `attempt_fit` method were trivial, inlined that into sole call site in the test body.